doc/services/storage/zms/zms.rst

```diff
@@ -0,0 +1,432 @@
+the first sector after garbage collecting it and erasing its content.
+
+Composition of a sector
+=======================
+A sector is organized in this form (example with 3 sectors):
+
+.. list-table::
+   :widths: 25 25 25
+   :header-rows: 1
+
+   * - Sector 0 (closed)
+     - Sector 1 (open)
+     - Sector 2 (empty)
+   * - Data_a0
+     - Data_b0
+     - Data_c0
+   * - Data_a1
+     - Data_b1
+     - Data_c1
+   * - Data_a2
+     - Data_b2
+     - Data_c2
+   * - GC_done
+     -    .
+     -    .
+   * -    .
+     -    .
+     -    .
+   * -    .
+     -    .
+     -    .
+   * -    .
+     - ATE_b2
+     - ATE_c2
+   * - ATE_a2
+     - ATE_b1
+     - ATE_c1
+   * - ATE_a1
+     - ATE_b0
+     - ATE_c0
+   * - ATE_a0
+     - GC_done
+     - GC_done
+   * - Close (cyc=1)
+     - Close (cyc=1)
+     - Close (cyc=1)
+   * - Empty (cyc=1)
+     - Empty (cyc=2)
+     - Empty (cyc=2)
+
+Definition of each element in the sector
+========================================
+
+``Empty ATE:`` is written when erasing a sector (last position of the sector).
+
+``Close ATE:`` is written when closing a sector (second to last position of the sector).
+
+``GC_done ATE:`` is written to indicate that the next sector has been already garbage
+collected. This ATE could be in any position of the sector.
+
+``ID-ATE:`` are entries that contain a 32 bits Key and describe where the data is stored, its
+size and its crc32
+
+``Data:`` is the actual value associated to the ID-ATE
+
+How does ZMS work?
+******************
+
+Mounting the Storage system
+===========================
+
+Mounting the storage starts by getting the flash parameters, checking that the file system
+properties are correct (sector_size, sector_count ...) then calling the zms_init function to
+make the storage ready.
+
+To mount the filesystem some elements in the zms_fs structure must be initialized.
+
+.. code-block:: c
+
+	struct zms_fs {
+		/** File system offset in flash **/
+		off_t offset;
+
+		/** Storage system is split into sectors, each sector size must be multiple of
+		 * erase-blocks if the device has erase capabilities
+		 */
+		uint32_t sector_size;
+		/** Number of sectors in the file system */
+		uint32_t sector_count;
+
+		/** Flash device runtime structure */
+		const struct device *flash_device;
+	};
+
+Initialization
+==============
+
+As ZMS has a fast-forward write mechanism, we must find the last sector and the last pointer of
+the entry where it stopped the last time.
+It must look for a closed sector followed by an open one, then within the open sector, it finds
+(recover) the last written ATE (Allocation Table Entry).
+After that, it checks that the sector after this one is empty, or it will erase it.
+
+ZMS ID-Data write
+===================
+
+To avoid rewriting the same data with the same ID again, it must look in all the sectors if the
+same ID exist then compares its data, if the data is identical no write is performed.
```

Comment:
Yes I was planning to do that because it optimizes the performance a lot in this case.
But I will do it in a separate PR as this is unrelated to this fixup

Reply:
Sure. Will be nice to have. (Also I don't know if it would make sense to have it as a runtime or ZMS instance parameter so that e.g. different callers can use it or not? At least I know that for the secure storage subsystem this better be disabled, but I am thinking of potential other uses of ZMS in parallel of that.)